### PR TITLE
7570 - Fix usage of "type" attribute in Addon.UI

### PR DESCRIPTION
**Pull request by DanielWilczek**

### CODE REVIEW

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs [hunk 1]
```diff
@@ -20,6 +19,12 @@ namespace G1ANT.Addon.UI
 {
     public class XPathUIElementBuilder : IXPathBuilder<object>
     {
+        public enum UiAutomationElement
+        {
+            Id = 0,
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: remove values assignment, it's unnecessary

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs [hunk 2]
```diff
@@ -160,14 +178,18 @@ public object Axis(XPathAxis xpathAxis, System.Xml.XPath.XPathNodeType nodeType,
             }
             if (xpathAxis == XPathAxis.Attribute)
             {
-                if (name == "id")
+                ControlType.Button.GetType();
+
+                if (name.ToLower() == "id")
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: ToLower can be done once

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 3]
```diff
@@ -18,6 +18,7 @@ public override void Initialize(IMainForm mainForm)
         {
             base.Initialize(mainForm);
             InitRootElement();
+            ControlType.Button.GetType();
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: It will not be called when the window tree panel is closed, move it to XPathUIElementBuilder or UIElement static constructor

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 4]
```diff
@@ -39,31 +40,38 @@ private void controlsTree_AfterCollapse(object sender, TreeViewEventArgs e)
...
+
+        private string GetTreeNodeName(AutomationElement element, int index)
         {
             if (element == null)
                 return "";
             string id = "";
             if (string.IsNullOrWhiteSpace(element.Current.AutomationId) == false)
                 id = $" #{element.Current.AutomationId}";
-            return $"{element.Current.LocalizedControlType}{id} \"{element.Current.Name}\"";
+            return $"{CutControlType(element.Current.ControlType.ProgrammaticName)}[{index}]{id} \"{element.Current.Name}\"";
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: remove index from the name of the node

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 5]
```diff
@@ -39,31 +40,38 @@ private void controlsTree_AfterCollapse(object sender, TreeViewEventArgs e)
...
-            if (string.IsNullOrWhiteSpace(element.Current.ClassName) == false)
-                result.AppendLine($"ClassName: {element.Current.ClassName}");
-            if (string.IsNullOrWhiteSpace(element.Current.Name) == false)
-                result.AppendLine($"Name: {element.Current.Name}");
-
+            if (!string.IsNullOrWhiteSpace(element.Current.AutomationId))
+                result.AppendLine($"id: {element.Current.AutomationId}");
+            if (element.Current.ControlType != null)
+                result.AppendLine($"type: {CutControlType(element.Current.ControlType.ProgrammaticName)}");
+            if (element.Current.ControlType != null)
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: duplicated "if" condition

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs [hunk 6]
```diff
@@ -160,14 +178,18 @@ public object Axis(XPathAxis xpathAxis, System.Xml.XPath.XPathNodeType nodeType,
             }
             if (xpathAxis == XPathAxis.Attribute)
             {
-                if (name == "id")
+                ControlType.Button.GetType();
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: move to static constructor

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs [hunk 7]
```diff
@@ -112,22 +117,36 @@ public object Operator(XPathOperator op, object left, object right)
         {
             if (op == XPathOperator.Eq)
             {
-                if (left is AutomationProperty property)
+                CompareFunc func = (elem, index) =>
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: split to separate lambda functions depends on "left" type, it has to be as fast as possible cause it is called many times

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs [hunk 8]
```diff
@@ -82,6 +87,7 @@ protected AutomationElement FindDescendantOrSelf(AutomationElement elem, Compare
         public AutomationElement Root { get; } = AutomationElement.RootElement;
         public XPathUIElementBuilder(AutomationElement root = null)
         {
+            ControlType.Button.GetType();
```
Comment (WiktoriaPrusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: why is that needed?

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 9]
```diff
@@ -121,5 +134,130 @@ private void refreshButton_Click(object sender, EventArgs e)
...
+                if (controlsTree.SelectedNode != null)
+                {
+                    if (controlsTree.SelectedNode.Tag is AutomationElement automationElement)
+                    {
+                        UIElement uiELement = new UIElement(automationElement);
+                        var element = UIElement.FromWPath(uiELement.ToWPath());
+                        if (element != null)
+                        {
+                            var window = GetTopLevelWindow(automationElement);
+                            if (window != null)
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: What should happen when window is hidden or null? Don't highlight control

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 10]
```diff
@@ -121,5 +134,130 @@ private void refreshButton_Click(object sender, EventArgs e)
...
+        
+        private Timer blinkTimer;
+        private void ClosingTimer_Tick(object sender, EventArgs e)
+        {
+            closingTimer.Enabled = false;
+            blinkTimer.Enabled = false;
+            rectangleForm.Close();
+        }
+
+        private void initializeRectangleForm(System.Windows.Rect rect)
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: Name should begin with uppercase

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 11]
```diff
@@ -121,5 +134,130 @@ private void refreshButton_Click(object sender, EventArgs e)
...
+        private Timer closingTimer;
+        private void RectangleForm_Shown(object sender, EventArgs e)
+        {
+            blinkTimer = new Timer();
+            blinkTimer.Interval = 300;
+            blinkTimer.Tick -= BlinkTimer_Tick;
+            blinkTimer.Tick += BlinkTimer_Tick;
+            blinkTimer.Enabled = true;
+
+            closingTimer = new Timer();
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: Do we really need second timer which will close blinking form? Can we do this in BlinkTimer_Tick after 5 "blinks"?

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs (hunk 8 above)
Comment (DanielWilczek) [reply G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: Microsoft fucked up initialization of static ControlType.
You need to use it at least use it once to initialize class.

Comment (pwitusik) [reply G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: @WiktoriaPrusik this is well known bug in UIAutomation, we need to call static constructor of ControlType on our own to allow this class instantation

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 12]
```diff
@@ -121,5 +134,130 @@ private void refreshButton_Click(object sender, EventArgs e)
...
+        /// <returns>The containing top-level window element.</returns>
+        private AutomationElement GetTopLevelWindow(AutomationElement element)
+        {
+            TreeWalker walker = TreeWalker.ControlViewWalker;
+            AutomationElement elementParent;
+            AutomationElement node = element;
+            do
+            {
+                elementParent = walker.GetParent(node);
+                if (elementParent == AutomationElement.RootElement) break;
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: Split to separate lines!

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 13]
```diff
@@ -121,5 +134,130 @@ private void refreshButton_Click(object sender, EventArgs e)
...
+                            if (window != null)
+                            {
+                                var iHandle = new IntPtr(window.Current.NativeWindowHandle);
+                                if (iHandle != IntPtr.Zero)
+                                {
+                                    RobotWin32.BringWindowToFront(iHandle);
+                                }
+                            }
+                            var rect = element.GetRectangle();
+                            if (rect != null) { initializeRectangleForm(rect); }
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: Split to separate lines!

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs [hunk 14]
```diff
@@ -121,5 +134,130 @@ private void refreshButton_Click(object sender, EventArgs e)
...
+            closingTimer.Tick += ClosingTimer_Tick;
+            closingTimer.Enabled = true;
+        }
+
+        private void BlinkTimer_Tick(object sender, EventArgs e)
+        {
+            rectangleForm.Visible = !rectangleForm.Visible;
+        }
+
+        private void controlsTree_MouseDown(object sender, MouseEventArgs e)
```
Comment (pwitusik) [inline G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: empty method!

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs (hunk 8 above)
Comment (WiktoriaPrusik) [reply G1ANT.Addon.UI/G1ANT.Addon.UI/Api/XPathUIElementBuilder.cs]: ok, thanks 👍

**Comment by DanielWilczek** [comment]
ping @pwitusik

File: G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs (hunk 9 above)
Comment (DanielWilczek) [reply G1ANT.Addon.UI/G1ANT.Addon.UI/Panels/UIControlsPanel.cs]: done
